### PR TITLE
spacy-transformers 1.3.5

### PR DESCRIPTION
**Pull request by skupr-anaconda**
**Destination channel:** main

### Links

- [PKG-2733](https://anaconda.atlassian.net/browse/PKG-2733) 
- [Upstream repository](https://github.com/explosion/spacy-transformers/tree/v1.3.5)
- Requirements:
  - https://github.com/explosion/spacy-transformers/blob/v1.3.5/build-constraints.txt
  - https://github.com/explosion/spacy-transformers/blob/v1.3.5/pyproject.toml
  - https://github.com/explosion/spacy-transformers/blob/v1.3.5/requirements.txt
  - https://github.com/explosion/spacy-transformers/blob/v1.3.5/setup.cfg
  - https://github.com/explosion/spacy-transformers/blob/v1.3.5/setup.py

### Explanation of changes:

- Clean up the recipe by conda-linter
- Add a compiler to `build`
- Add `cython` and `numpy` to `host`
- Update runtime pinnings, add missing `sentencepiece` and `protobuf` (extracting them from `transformers[sentencepiece]`), see https://github.com/explosion/spacy-transformers/blob/ba6bf8202662f59aa1862bd4c5dd6fd0119a3335/requirements.txt#L3-L3C28

### Notes:

- Updating because of the `pytorch 2.3.0` release

[PKG-2733]: https://anaconda.atlassian.net/browse/PKG-2733?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: recipe/meta.yaml [hunk 1]
```diff
@@ -1,56 +1,55 @@
...
   number: 0
-  # spacy isn't available on s390x
-  skip: True  # [py<36 or s390x]
-  # At present, there are no 3.11 compatible Pytorch version for win or ppc
-  # There is a bug requiring PyTorch 1.10 on osx-64 which is not available for 3.11
-  # As transformers 4.29.2 b0 has the same issue, we apply its skipping logic here  
-  skip: True  # [py>310 and (ppc64le or win or (osx and x86_64))]
-  script: {{ PYTHON }} -m pip install . --no-deps --no-build-isolation -vv
+  # spacy isn't available on s390x.
+  # transformers isn't available for py>=312 currently.
```
Comment (cbouss) [inline recipe/meta.yaml]: What does it take to build transformers for 3.12, have you tried?

File: recipe/meta.yaml [hunk 2]
```diff
@@ -1,56 +1,55 @@
...
-    - spacy >=3.5.0,<4.0.0
-    - spacy-alignments >=0.7.2,<1.0.0
+    - spacy >=3.5.0,<4.1.0
+    # transformers[sentencepiece] means 'extras["sentencepiece"] = deps_list("sentencepiece", "protobuf")',
+    # see https://github.com/explosion/spacy-transformers/blob/v1.3.5/requirements.txt#L3-L3C28 and
+    # https://github.com/huggingface/transformers/blob/345b9b1a6a308a1fa6559251eb33ead2211240ac/setup.py#L300-L300C65
+    # Anaconda set the transformers upper bound to <4.42.0 because of compatibility with 'tokenisers' versions
+    - transformers >=3.4.0,<4.42.0
+    - sentencepiece >=0.1.91,!=0.1.92
+    - protobuf
```
Comment (cbouss) [inline recipe/meta.yaml]: I don't see requirements.txt actually being used. We should follow https://github.com/explosion/spacy-transformers/blob/v1.3.5/setup.cfg#L32-L40
Perhaps what is in requirements.txt is useful for tests though.
```suggestion
    - transformers >=3.4.0,<4.37.0
```

File: recipe/meta.yaml [hunk 3]
```diff
@@ -1,56 +1,55 @@
...
+    - pytorch >=1.8.0
     - srsly >=2.4.0,<3.0.0
-    - transformers >=3.4.0,<4.30.0
+    - spacy-alignments >=0.7.2,<1.0.0
 
 test:
-  requires:
-    - pip
```
Comment (cbouss) [inline recipe/meta.yaml:46]: Any reason for removing tests?

File: recipe/meta.yaml (hunk 2 above)
Comment (skupr-anaconda) [reply recipe/meta.yaml]: It fails if we use `transformers >=3.4.0,<4.37.0` because the old version of transformers is picked up and then the incompatible version tokenizers will be used. I can update transformers but I haven't evaluated yet how much time it will take

File: recipe/meta.yaml (hunk 3 above)
Comment (skupr-anaconda) [reply recipe/meta.yaml:46]: It's because I fixed the upper bound of transformers. The maintainers are moving its upper bound every new release so the latest is also compatible. I constrained it by `<4.42.0` as we have `transformers 4.37.2` currently

File: recipe/meta.yaml (hunk 1 above)
Comment (skupr-anaconda) [reply recipe/meta.yaml]: Rebuilding transformers https://github.com/AnacondaRecipes/transformers-feedstock/pull/11

File: recipe/meta.yaml (hunk 2 above)
Comment (cbouss) [reply recipe/meta.yaml]: Ok, you can use this as a patch
https://github.com/explosion/spacy-transformers/pull/418.patch and re-enable pip check.
(The change is compatible: https://github.com/explosion/spacy-transformers/compare/v1.3.5..master)
But I would still remove sentencepiece and protobuf which I don't see being used in setup.cfg.

File: recipe/meta.yaml (hunk 1 above)
Comment (skupr-anaconda) [reply recipe/meta.yaml]: Done

File: recipe/meta.yaml (hunk 2 above)
Comment (skupr-anaconda) [reply recipe/meta.yaml]: Done. But there was an issue that `setup.cfg` on **PyPI** is not the same as on the **GitHub**. I temporarily replaced `source/url`

File: recipe/meta.yaml (hunk 3 above)
Comment (skupr-anaconda) [reply recipe/meta.yaml:46]: Fixed

**Comment by skupr-anaconda** [comment]
@cbouss spacy-transformers failed with python 3.12.4 and pydantic <2.7.4:
`TypeError: ForwardRef._evaluate() missing 1 required keyword-only argument: 'recursive_guard'`

See also https://github.com/pydantic/pydantic/issues/9637.
To solve this issue we have to update pydantic to >=2.7.4 and make from scratch a new feedstock pydantic-core 2.18.4, see https://github.com/pydantic/pydantic/blob/v2.7.4/pyproject.toml#L51.

I would skip py312 now because spacy-transformers will be needed only for the next release of the spacy language models https://metayaml-conda.fly.dev/metayaml/reverse_dependencies?name=spacy-transformers

**Comment by danpetry** [comment]
I got a download count of 1132 for spacy-model-en_core_web_trf in May from defaults. 4956 for spacy-transformers. Not very many. Also to update spacy-transformers "properly" would probably involve removing the upper bound and updating those other models that depend on spacy-transformers. Seems not worth it, we could skip the update of spacy-transformers completely, or yeah just skip py 3.12 since there's quite some work been done already

**Comment by danpetry** [comment]
the current version will "just work" with new pytorch according to the current pinning?

File: recipe/meta.yaml [hunk 4]
```diff
@@ -1,56 +1,65 @@
...
-    # TODO: Working around osx bug importing two openmp variants.
-    - pytorch >=1.8.0     # [not (osx and x86_64)]
-    - pytorch >1.9,<1.12  # [osx and x86_64]
-    - spacy >=3.5.0,<4.0.0
-    - spacy-alignments >=0.7.2,<1.0.0
+    - spacy >=3.5.0,<4.1.0
+    # Anaconda applies a patch and sets the transformers upper bound to <4.42.0 because of compatibility with 'tokenisers' versions
+    - transformers >=3.4.0,<4.42.0
+    - pytorch >=1.8.0
     - srsly >=2.4.0,<3.0.0
```
Comment (lorepirri) [inline recipe/meta.yaml:45]:
```suggestion
    - srsly >=2.4.0,<3.0.0
    - dataclasses >=0.6  # [py<37]
```

Comment (lorepirri) [reply recipe/meta.yaml:45]: because
- `dataclasses>=0.6,<1.0; python_version < "3.7"`
- and we don't skip `py<37`

File: recipe/meta.yaml [hunk 5]
```diff
@@ -1,56 +1,65 @@
...
+{% set version = "1.3.5" %}
 
 package:
   name: {{ name }}
   version: {{ version }}
 
 source:
-  url: https://pypi.io/packages/source/{{ name[0] }}/{{ name }}/{{ name }}-{{ version }}.tar.gz
-  sha256: 859c2093fad9ff41005bc55c0018d4290bdd62454f76bddbcf318aae85f307b5
+  # To apply fix-transformers-upper-bound.patch we need the same setup.cfg as on the GitHub but this file is different on PyPI.
```
Comment (lorepirri) [inline recipe/meta.yaml:10]: for me it's fine this way,
but to me it looks like the portion of the patch is the same and the `patch` command should be able to apply the patch even if the rest of the file is different

https://inspector.pypi.io/project/spacy-transformers/1.3.5/packages/3f/63/5c76fdfa6bc4905c41031ab7ef2a5da221e438141e629ddf3f2729fd778c/spacy_transformers-1.3.5.tar.gz/spacy_transformers-1.3.5/setup.cfg#line.32

**Review by lorepirri — APPROVED** [review]
LGTM, just a couple of nitpicks.

I suppose you changed `doc_url` to the repo because that link to the docs is a bit generic, right?

File: recipe/meta.yaml (hunk 4 above)
Comment (skupr-anaconda) [reply recipe/meta.yaml:45]: Fixed

File: recipe/meta.yaml (hunk 5 above)
Comment (skupr-anaconda) [reply recipe/meta.yaml:10]: Applying the patch failed. That's why it's not possible to use the tarball from PyPI

**Comment by skupr-anaconda** [comment]
> I got a download count of 1132 for spacy-model-en_core_web_trf in May from defaults. 4956 for spacy-transformers. Not very many. Also to update spacy-transformers "properly" would probably involve removing the upper bound and updating those other models that depend on spacy-transformers. Seems not worth it, we could skip the update of spacy-transformers completely, or yeah just skip py 3.12 since there's quite some work been done already

@danpetry I would prefer uploading it without py<312

File: recipe/meta.yaml (hunk 5 above)
Comment (lorepirri) [reply recipe/meta.yaml:10]: All good

**Comment by ifitchet** [comment]
Of interest, the GH repo says the default branch is `main` but `.gitmodules` uses `master`.

**Comment by skupr-anaconda** [comment]
> Of interest, the GH repo says the default branch is `main` but `.gitmodules` uses `master`.

I think someone incorrectly set the `main` branch as default in the feedstock settings

File: recipe/meta.yaml [hunk 6]
```diff
@@ -71,5 +81,3 @@ extra:
     - honnibal
     - ines
     - adrianeboyd
-  skip-lints:
-    - cython_needs_compiler
```
Comment (ifitchet) [inline recipe/meta.yaml]: Why did you delete this when the linter still complains?
(I assume it is a linter feature that it complains that the C++ compiler is not the (expected) C compiler.)

**Comment by lorepirri** [comment]
> > Of interest, the GH repo says the default branch is `main` but `.gitmodules` uses `master`.
> 
> I think someone incorrectly set the `main` branch as default in the feedstock settings

@skupr-anaconda I set the def branch to `master`, because that's what on gitmodules and this PR merges against master.

@ifitchet nice catch!

File: recipe/meta.yaml (hunk 6 above)
Comment (skupr-anaconda) [reply recipe/meta.yaml]: It's because my first try was opening the PR against the main branch.

Comment (skupr-anaconda) [reply recipe/meta.yaml]: Fixed

**Comment by lorepirri** [comment]
@ifitchet @boldorider4 @rafaelmartins anyone up for a review on this one?

**Review by rafaelmartins — APPROVED** [review]
